webの基礎/no.7/answer.md

```diff
@@ -0,0 +1,277 @@
+
+```
+etag: "nzzXVFNxmk/Q8YsfYIB7ww=="
+expires: Wed, 26 Jan 2022 05:12:47 GMT
+last-modified: Sat, 23 Jan 2021 01:35:24 GMT
+```
+
+`ETag`と`last-modified`が設定されていることから`expires`の期限が切れると条件付きの GET(`If-None-Match`と`If-Modified-Since`)が行われることが分かります。
+
+画像は以下のようなレスポンスになっており
+
+```
+cache-control: max-age=604800, must-revalidate
+last-modified: Thu, 21 Jun 2018 15:18:06 GMT
+```
+
+キャッシュの期限が切れたら、必ずオリジンのサーバーへリソースの確認を行うようになっていることが分かります。
+
+- [Facebook](https://www.facebook.com/)
+
+![facebookスクリーンショット](images/facebook.png)
+
+スクリーンショットに載せている HTML を確認すると以下のようなレスポンスになっており
+
+```
+cache-control: private, no-cache, no-store, must-revalidate
+expires: Sat, 01 Jan 2000 00:00:00 GMT
+pragma: no-cache
+```
+
+`no-store`を指定していることから、twitter と同様にどのキャッシュにも保存しないようにしていることが分かります。`pragma`が残っているのは互換性担保？
+
+JS と CSS のレスポンスは以下のようになっており
+
+```
+cache-control: public,max-age=31536000,immutable
+expires: Sun, 30 Jan 2022 00:24:48 GMT
+last-modified: Mon, 01 Jan 2001 08:00:00 GMT
+```
+
+`cache-control`の`max-age`と`expires`でキャッシュの有効期限を設定しており、リロード時にもローカルのキャッシュを使えるように`immutable`が指定してあることが分かります。
+
+画像のレスポンスは以下のようになっており
+
+```
+cache-control: max-age=1209600, no-transform
+last-modified: Tue, 24 Dec 2019 13:15:59 GMT
+```
+
+`max-age`を用いてキャッシュの有効期限を設定しており、`no-transform`を用いることで画像の圧縮を防いでいることが分かります。
+
+## 課題２（実装）
+
+http-cache-sample-app にアプリケーションを実装しました。
+1 日キャッシュする画像、キャッシュしない画像(`no-store`)、条件付き Get を行う画像を配信するエンドポイントを用意し、レスポンスの結果が以下の画像になります。
+![Chromeでの確認のスクリーンショット](images/sample.png)
+
+- 1 日キャッシュする画像
+
+![cacheスクリーンショット](images/cache.png)
+
+- キャッシュしない画像
+
+![no-cacheスクリーンショット](images/no-cache.png)
+
+- 条件付き Get を行う画像
+
+![condition-getスクリーンショット](images/condition-get.png)
+
+## 課題３（成果物に関する質問）
+
+ブラウザキャッシュを使うべきではないケースを３つ考えて、ペアと会話してみてください。どんなサービスの？どんなページで？どんなファイルを？キャッシュしてはいけないのでしょうか？
+
+- リアルタイムに動的なコンテンツを取り扱うページやコンテンツ
+  ブラウザのキャッシュは、各ユーザー側にキャッシュがあるため、コントロールが聞きにくいので頻繁にキャッシュするリソースが変わるページやコンテンツではブラウザのキャッシュを有効活用することが出来ません。
+
+  参考：https://qiita.com/inagakkie/items/f615d7d8cf172eb15a4b#%E3%82%B3%E3%83%B3%E3%83%86%E3%83%B3%E3%83%84%E3%82%AD%E3%83%A3%E3%83%83%E3%82%B7%E3%83%A5%E6%88%A6%E7%95%A5
+
+- 大量の画像などのコンテンツを扱い、且つユーザーが複数ページを回遊するようなサービス(ニュースメディアとか？)
+  ブラウザがキャッシュ出来るコンテンツの量は限度があります。静的コンテンツが大量にあり、且つユーザーが複数ページを回遊した場合、ブラウザのキャッシュが削除されてしまう可能性があるため、キャッシュを有効活用することが出来ません。逆に共有キャッシュであれば、有効活用出来そう？(メンターに聞いてみたい)
+
+- 個人情報を取り扱うページやファイル(※ブラウザキャッシュで気を付ける例)
+  ブラウザのキャッシュはブラウザを閉じたとしてもデータが残っていることがあるため、マルウェアへの感染や PC の盗難などによって、ブラウザに残っている個人情報を盗まれてしまう可能性があります。
+
+  参考：https://www.lifehacker.jp/2013/06/130628caching.html
+
+- 個人情報を取り扱うページやファイル(※ブラウザキャッシュではなく、共有キャッシュで気を付ける例)
+  HTTP キャッシュの仕組みは非常に複雑で、ちょっとした設定ミスによって情報漏洩に繋がる恐れがあるため、個人情報などのセンシティブな情報をキャッシュするべきではありません。メルカリや SOD などの大きな会社でもキャッシュの設定ミスによるインシデントが発生しています。([メルカリ](https://engineering.mercari.com/blog/entry/2017-06-22-204500/)[SOD](https://scan.netsecurity.ne.jp/article/2020/03/31/43887.html))
+
+## 課題４（クイズ）
+
+No.1: Cache-Control ヘッダーの`no-cache`と`no-store`の違いを説明してください。
+No.2: `ETag`と`last-modified`を指定したとき、どちらが優先されるでしょうか？
```

Comment:
No2の回答です。

ETagです。（厳密に言うと、If-Modified-Since と一緒に使用した場合に、 If-None-Match が優先される？）ETagの検証の方が精度が高いとされているからですかね。

Reply:
合っています！


中野の回答は以下です！
If-Modified-SinceよりもIf-None-Matchが優先されるため、EtagとLast-Modifiedをレスポンスヘッダーに設定した場合、Etagが優先されます。

[https://triple-underscore.github.io/RFC7232-ja.html#section-6](https://triple-underscore.github.io/RFC7232-ja.html#section-6)

[https://blog.redbox.ne.jp/http-header-tuning.html#:~:text=ETAGはLast-Modifiedより,Matchが優先されます。](https://blog.redbox.ne.jp/http-header-tuning.html#:~:text=ETAG%E3%81%AFLast%2DModified%E3%82%88%E3%82%8A,Match%E3%81%8C%E5%84%AA%E5%85%88%E3%81%95%E3%82%8C%E3%81%BE%E3%81%99%E3%80%82)

[https://developer.mozilla.org/ja/docs/Web/HTTP/Headers/If-None-Match](https://developer.mozilla.org/ja/docs/Web/HTTP/Headers/If-None-Match)